### PR TITLE
feat(webhook): Single-identity mTLS webhook configuration

### PR DESCRIPTION
**Pull request by jcavanagh**
Allows configuring a single X509 identity to use as the client identity for all outgoing webhooks.
Also enhances existing trust functionality to accept PEM CA certificates.

When using keystores, an encrypted private key entry is required - passwordless unencrypted private keys in keystores are not supported.  Use PEM instead.
Similarly, keystores without passwords are also not supported.

When using PEM, assumes no key encryption.

### CODE REVIEW

File: orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java [hunk 1]
```diff
@@ -105,9 +109,14 @@ public ClientHttpRequestFactory webhookRequestFactory(
                   }
 
                   return response;
-                })
-            .build();
-    OkHttp3ClientHttpRequestFactory requestFactory = new OkHttp3ClientHttpRequestFactory(client);
+                });
+
+    if (webhookProperties.isInsecureSkipHostnameVerification()) {
+      builder.hostnameVerifier((hostname, session) -> true);
```
Comment (jcavanagh) [inline orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java:115]: Hostname verification needs to be disabled for testing, as OkHttp does validate certificate hosts by default

File: orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java [hunk 2]
```diff
@@ -73,9 +76,9 @@ public WebhookConfiguration(WebhookProperties webhookProperties) {
   @Bean
   @ConditionalOnMissingBean(RestTemplate.class)
   public RestTemplate restTemplate(ClientHttpRequestFactory webhookRequestFactory) {
-    RestTemplate restTemplate = new RestTemplate(webhookRequestFactory);
+    var restTemplate = new RestTemplate(webhookRequestFactory);
```
Comment (jcavanagh) [inline orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java:79]: Converted a bunch of types to `var` for legibility

File: orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java [hunk 3]
```diff
@@ -116,58 +125,123 @@ public ClientHttpRequestFactory webhookRequestFactory(
...
   private X509TrustManager webhookX509TrustManager() {
-    List<X509TrustManager> trustManagers = new ArrayList<>();
+    var trustManagers = new ArrayList<X509TrustManager>();
 
     trustManagers.add(getTrustManager(null));
-    getCustomKeyStore().ifPresent(keyStore -> trustManagers.add(getTrustManager(keyStore)));
+    getCustomTrustStore().ifPresent(keyStore -> trustManagers.add(getTrustManager(keyStore)));
+
+    if (webhookProperties.isInsecureTrustSelfSigned()) {
+      trustManagers.add(
```
Comment (jcavanagh) [inline orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java:134]: Another testing-only option to disable trust checks for situations that cannot or do not provide custom CA files.

File: orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java [hunk 4]
```diff
@@ -116,58 +125,123 @@ public ClientHttpRequestFactory webhookRequestFactory(
...
     }
 
-    KeyStore keyStore;
-    try {
-      keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
-    } catch (KeyStoreException e) {
-      throw new RuntimeException(e);
+    if (StringUtils.isNotEmpty(identitySettings.getIdentityStore())) {
+      var identity =
+          X509IdentitySource.fromKeyStore(
```
Comment (jcavanagh) [inline orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java:188]: Constructs a client identity from a keystore.  The keystore must be password protected, and the private key entry in the store also must be password protected.  For unencrypted keys, use PEM config instead.

File: orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java [hunk 5]
```diff
@@ -116,58 +125,123 @@ public ClientHttpRequestFactory webhookRequestFactory(
...
+                return password == null ? new char[0] : password.toCharArray();
+              },
+              () -> {
+                var password = identitySettings.getIdentityStoreKeyPassword();
+                return password == null ? new char[0] : password.toCharArray();
+              });
+      return Optional.of(identity.load());
+    } else if (StringUtils.isNotEmpty(identitySettings.getIdentityKeyPem())) {
+      return Optional.of(
+          X509IdentitySource.fromPEM(
```
Comment (jcavanagh) [inline orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java:202]: Constructs a client identity from PEM key and PEM cert.  No encryption or password-protection supported.

File: orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java [hunk 6]
```diff
@@ -116,58 +125,123 @@ public ClientHttpRequestFactory webhookRequestFactory(
...
+
+  private Optional<KeyStore> getCustomTrustStore() {
+    var trustSettings = webhookProperties.getTrust();
+    if (trustSettings == null || !trustSettings.isEnabled()) {
+      return Optional.empty();
+    }
+
+    // Use keystore first if set, then try PEM
+    if (StringUtils.isNotEmpty(trustSettings.getTrustStore())) {
+      KeyStore keyStore;
```
Comment (jcavanagh) [inline orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java:219]: Loads CA trust material from a keystore (preexisting code)

File: orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java [hunk 7]
```diff
@@ -116,58 +125,123 @@ public ClientHttpRequestFactory webhookRequestFactory(
...
+      try (FileInputStream file = new FileInputStream(trustSettings.getTrustStore())) {
+        keyStore.load(file, trustSettings.getTrustStorePassword().toCharArray());
+      } catch (CertificateException | IOException | NoSuchAlgorithmException e) {
+        throw new RuntimeException(e);
+      }
+
+      return Optional.of(keyStore);
+    } else if (StringUtils.isNotEmpty(trustSettings.getTrustStorePem())) {
+      try {
+        return Optional.of(TrustStores.loadPEM(Path.of(trustSettings.getTrustStorePem())));
```
Comment (jcavanagh) [inline orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java]: Loads CA trust material from a PEM certificate list (new code)

File: orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java [hunk 8]
```diff
@@ -116,58 +125,123 @@ public ClientHttpRequestFactory webhookRequestFactory(
...
+      try (FileInputStream file = new FileInputStream(trustSettings.getTrustStore())) {
+        keyStore.load(file, trustSettings.getTrustStorePassword().toCharArray());
+      } catch (CertificateException | IOException | NoSuchAlgorithmException e) {
+        throw new RuntimeException(e);
+      }
+
+      return Optional.of(keyStore);
+    } else if (StringUtils.isNotEmpty(trustSettings.getTrustPem())) {
+      try {
+        return Optional.of(TrustStores.loadPEM(Path.of(trustSettings.getTrustPem())));
```
Comment (jcavanagh) [inline orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookConfiguration.java:235]: Loads CA trust material from a PEM certificate list (new code)

File: orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookProperties.java [hunk 9]
```diff
@@ -53,18 +53,39 @@ public class WebhookProperties {
...
+
   @Data
   @NoArgsConstructor
   public static class TrustSettings {
     private boolean enabled;
     private String trustStore;
     private String trustStorePassword;
+    // Default as JKS instead of PKCS12 for backward compatibility
+    private String trustStoreType = "JKS";
+    private String trustPem;
```
Comment (dbyron-sf) [inline orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/config/WebhookProperties.java:75]: For the record, a curl command like
```
$ curl --cert identityCertPem --key identityKeyPem --cacert trustPem https://somewhere
```
corresponds to these pem properties.

**Review by dbyron-sf — APPROVED** [review]

**Pull request merged**